src/server/apsystem.ts

```diff
@@ -192,7 +221,86 @@ export default class ActivityPubSystem {
     })
   }
 
-  async sendTo (actorURL: string, fromActor: string, activity: APActivity): Promise<void> {
+  async get (fromActor: string, urlOrObject: ObjectField): Promise<APObject> {
+    if (typeof urlOrObject === 'string') {
+      const response = await this.signedFetch(fromActor, {
+        method: 'GET',
+        url: urlOrObject,
+        headers: {}
+      })
+      if (!response.ok) {
+        throw createError(
+          500,
+          `Cannot fetch data for ${fromActor} at ${urlOrObject}: http status ${response.status} - ${await response.text()}`
+        )
+      }
+
+      return await response.json()
+    } else {
+      return urlOrObject
+    }
+  }
+
+  async * getAll (
+    fromActor: string,
+    items: ObjectField[]
+  ): AsyncIterableIterator<APObject> {
+    for (const itemOrUrl of items) {
+      const item = await this.get(fromActor, itemOrUrl)
+
+      yield item
+    }
+  }
+
+  async * iterateCollection (
```

Comment:
one thing i found while implementing replies iteration is that mastodon will embed a collection on the replies attribute which includes some replies, but also points to next page even when it's empty, so if you don't check properly you'll end up with an infinite loop. i couldn't find the source code for mastodon's implementation so i had to look ahead and stop if the next page id is equal to the current one. it would've been great to just validate totalItems but it's not always the case.

just a long winded way of asking if you found this case and if it applies to this function :P

Reply:
That's a good point! I'll add a check and break if the collection page is empty